### ISSUE TITLE
IDP configuration

### ISSUE THREAD

**Issue by Sleepy-GH** [minio/console#670]
Hey folks,

So I'm trying to set up Console and Minio through the operator and have them both pointing at Keycloak for authentication.
Are there any instructions for how to configure Keycloak? I'm particularly having issues with the scopes. When I don't add any specific scopes I get the following error in Keycloak:
```bash
13:41:56,419 ERROR [org.keycloak.services] (default task-228) KC-SERVICES0093: Invalid parameter value for: scope
13:41:56,420 WARN  [org.keycloak.events] (default task-228) type=LOGIN_ERROR, realmId=<realm>, clientId=minio, userId=null, ipAddress=xxx, error=invalid_request, response_type=code, redirect_uri=https://<my-keycloak>/oauth_callback, response_mode=query
```
I did some looking around and I noticed that Console expects the following extra scoped to be configured:
- user_metadata
- app_metadata
- openid

So, I added these with no particular mappers. This got rid of all the error messages when trying to log in, but it keeps me in a loop where I log in and just get redirected to the login screen again.

Does anyone have an idea what I'm missing?

**Comment by Sleepy-GH** [comment]
Actually, I just double checked the logs of Console and found this:
```bash
2021-03-26 15:45:04.935631 I | 2021/03/26 15:45:04 error.go:44: original error: invalid Login
2021-03-26 15:45:04.935790 I | 2021/03/26 15:45:04 error.go:124: debugging error: Post "https://<my-keycloak>/auth/realms/Kubernetes/protocol/openid-connect/token": x509: certificate signed by unknown authority
```

So, I do have a self-signed certificate, but I also have my certificate configured in the Tenant spec:
```yaml
spec:
...
  console:
    ...
    # Keycloak CA cert because self-signed
    externalCaCertSecret:
      - name: keycloak-ca
     ...
```
When I didn't have that configured, I had issues earlier on in the login process (as in, the login screen only showed an error)

**Comment by Alevsk** [comment]
> Hey folks,
> 
> So I'm trying to set up Console and Minio through the operator and have them both pointing at Keycloak for authentication.
> Are there any instructions for how to configure Keycloak? I'm particularly having issues with the scopes. When I don't add any specific scopes I get the following error in Keycloak:
> 
> ```shell
> 13:41:56,419 ERROR [org.keycloak.services] (default task-228) KC-SERVICES0093: Invalid parameter value for: scope
> 13:41:56,420 WARN  [org.keycloak.events] (default task-228) type=LOGIN_ERROR, realmId=<realm>, clientId=minio, userId=null, ipAddress=xxx, error=invalid_request, response_type=code, redirect_uri=https://<my-keycloak>/oauth_callback, response_mode=query
> ```
> 
> I did some looking around and I noticed that Console expects the following extra scoped to be configured:
> 
> * user_metadata
> * app_metadata
> * openid
> 
> So, I added these with no particular mappers. This got rid of all the error messages when trying to log in, but it keeps me in a loop where I log in and just get redirected to the login screen again.
> 
> Does anyone have an idea what I'm missing?

Hi, which version of console are you using? can you try with master? we eliminate the requirement of user_metadata and app_metadata and instead we only ask for `openid,profile,email` which is pretty standard

**Comment by Alevsk** [comment]
> Actually, I just double checked the logs of Console and found this:
> 
> ```shell
> 2021-03-26 15:45:04.935631 I | 2021/03/26 15:45:04 error.go:44: original error: invalid Login
> 2021-03-26 15:45:04.935790 I | 2021/03/26 15:45:04 error.go:124: debugging error: Post "https://<my-keycloak>/auth/realms/Kubernetes/protocol/openid-connect/token": x509: certificate signed by unknown authority
> ```
> 
> So, I do have a self-signed certificate, but I also have my certificate configured in the Tenant spec:
> 
> ```yaml
> spec:
> ...
>   console:
>     ...
>     # Keycloak CA cert because self-signed
>     externalCaCertSecret:
>       - name: keycloak-ca
>      ...
> ```
> 
> When I didn't have that configured, I had issues earlier on in the login process (as in, the login screen only showed an error)

Make sure to add this certificate to MinIO itself too:

```
spec:
    externalCaCertSecret:
       - name: keycloak-ca
...
  console:
    ...
    # Keycloak CA cert because self-signed
    externalCaCertSecret:
      - name: keycloak-ca
     ...
```

**Comment by Sleepy-GH** [comment]
Thanks for looking into this. I'm using v0.6.3.

It might be a bit before I can test this. It's a busy period currently, but I definitely will give it a shot when time permits!

**Comment by jgato** [comment]
@Sleepy-GH could you solve your issue. I am also trying to enable keycloak authentication but not lucky. 

https://github.com/minio/operator/issues/603

Now I am trying even disable every ssl option to find out the problem.

**Comment by Sleepy-GH** [comment]
@Alevsk Is there a container tag published with the changes that I can configure?

**Comment by Sleepy-GH** [comment]
So I did some debugging and here's how far I got.

I configured a test tenant with the details below:
```yaml
apiVersion: minio.min.io/v2
kind: Tenant
metadata:
  name: test
  labels:
    app: minio-test
  annotations:
    prometheus.io/path: /minio/prometheus/metrics
    prometheus.io/port: "9000"
    prometheus.io/scrape: "true"
spec:
  image: private.docker.registry/docker-hub/minio/minio:RELEASE.2021-03-01T04-20-55Z
  imagePullPolicy: IfNotPresent

  ## Secret with credentials to be used by MinIO Tenant.
  ## Refers to the secret object created above.
  credsSecret:
    name: minio

  # Keycloak CA cert because self-signed
  externalCaCertSecret:
    - name: keycloak-ca

  ## Specification for MinIO Pool(s) in this Tenant.
  pools:
    - servers: 4
      volumesPerServer: 1
      volumeClaimTemplate:
        metadata:
          name: data
        spec:
          storageClassName: openebs-hostpath
          accessModes:
            - ReadWriteOnce
          resources:
            requests:
              storage: 4Gi

  ## Mount path where PV will be mounted inside container(s).
  mountPath: /export

  requestAutoCert: false

  s3:
    bucketDNS: false

  podManagementPolicy: Parallel

  # Env vars for Minio Pods
  env:
  - name: MINIO_IDENTITY_OPENID_CONFIG_URL
    value: "https://keycloak.<my_dns>/auth/realms/Kubernetes/.well-known/openid-configuration" # Your external identity provide configuration
  - name: MINIO_IDENTITY_OPENID_CLIENT_ID
    value: "minio"
  - name: MINIO_IDENTITY_OPENID_SCOPES
    value: "openid,profile,email"
  - name: MINIO_IDENTITY_OPENID_CLAIM_NAME
    value: "minio_policy"

  serviceMetadata:
    minioServiceLabels:
      label: minio-test-svc
    minioServiceAnnotations:
      v2.min.io: minio-test-svc
    consoleServiceLabels:
      label: console-test-svc
    consoleServiceAnnotations:
      v2.min.io: console-test-svc

  console:
    image: private.docker.registry/docker-hub/minio/console:v0.6.8
    imagePullPolicy: "IfNotPresent"
    replicas: 1
    consoleSecret:
      name: console-secret
    # Keycloak CA cert because self-signed
    externalCaCertSecret:
      - name: keycloak-ca
    securityContext:
      runAsUser: 1000
      runAsGroup: 2000
      runAsNonRoot: true
    env:
    - name: CONSOLE_IDP_URL
      value: "https://keycloak.<my_dns>/auth/realms/Kubernetes" # Your external identity provide URL
    - name: CONSOLE_IDP_CLIENT_ID
      value: "minio"
    - name: CONSOLE_IDP_SECRET
      valueFrom:
        secretKeyRef:
          name: console-idp-secret
          key: console_secret
    - name: CONSOLE_IDP_CALLBACK
      value: "https://console-test.<my_dns>/oauth_callback"
```
Creation of the deployment all goes fine. 

With the correct certificate in the keycloak-ca secret I get the loading screen with the "Login With SSO" button. When I push this button I get redirected to Keycloak, I log in and I get redirected to the same login screen. When I then check the server logs I find the following entries:
```
2021-05-04 17:49:53.889566 I | 2021/05/04 17:49:53 error.go:46: original error: invalid Login
2021-05-04 17:49:53.889669 I | 2021/05/04 17:49:53 error.go:134: debugging error: Post "https://keycloak.<my_dns>/auth/realms/Kubernetes/protocol/openid-connect/token": x509: certificate signed by unknown authority
```

When I change the keycloak-ca secret to contain a cert that is totally unrelated I get different behaviour. When I then try and go to the login screen I'm instantly presented with an error screen and a different error log message like this:
```
2021-05-04 17:37:03.627159 I | 2021/05/04 17:37:03 error.go:46: original error: Get "https://keycloak.<my_dns>/auth/realms/Kubernetes/.well-known/openid-configuration": x509: certificate signed by unknown authority (possibly because of "crypto/rsa: verification error" while trying to verify candidate authority certificate "kube-ca")
```

So in some part handling of self-signed certs does not fully work I think.
Hope this helps.

**Comment by jgato** [comment]
Well, you at least have an error. I guess these error came from the Keycloack logs, right? How is your configured your ingress? To which service/port are you pointing?
Do you know/understand why the keycloak configuration is set some variables in the tenant specification but others in the console specification? I dont know if you have seen this other bug I opened:
https://github.com/minio/operator/issues/620
I cannot access the console with that error.

**Comment by itmwiw** [comment]
Hello,
I have the same problem. In my case, the external CA is ingress-operator. I added it to tenant and console spec using:
    externalCaCertSecret:
      - name: keycloak-ca
I've checked to be sure that all my pods (console and tenant) have the CA in /tmp/certs/CAs/ca-0.crt 
I also added the same CA o my chrome truststore to be sure that my keycloak it's the correct CA and that was OK too.
I'm stuck: I still have the same message :
 2021-05-30 22:19:15.046946 I | 2021/05/30 22:19:15 error.go:154: debugging error: Post "https://xxx.xxx.com/auth/realms/keycloak-xxxt/protocol/openid-connect/token": x509: certificate signed by unknown authority
 tested on console version 0.7.1 and 0.7.4
